hook-common/src/pgqueue.rs

```diff
@@ -284,30 +316,93 @@ impl<J: std::marker::Send, M: std::marker::Send> PgQueueJob for PgJob<J, M> {
 #[derive(Debug)]
 pub struct PgTransactionJob<'c, J, M> {
     pub job: Job<J, M>,
-    pub transaction: sqlx::Transaction<'c, sqlx::postgres::Postgres>,
+
+    /// The open transaction this job came from. If multiple jobs were queried at once, then this
+    /// transaction will be shared between them (across async tasks and threads as necessary). See
+    /// below for more information.
+    shared_txn: Arc<Mutex<SharedTxn<'c>>>,
+}
+
+// Container struct for a batch of PgTransactionJob. Includes a reference to the shared transaction
+// for committing the work when all of the jobs are finished.
+pub struct PgTransactionBatch<'c, J, M> {
+    pub jobs: Vec<PgTransactionJob<'c, J, M>>,
+
+    /// The open transaction the jobs in the Vec came from. This should be used to commit or
+    /// rollback when all of the work is finished.
+    shared_txn: Arc<Mutex<SharedTxn<'c>>>,
+}
+
+impl<'c, J, M> PgTransactionBatch<'_, J, M> {
+    pub async fn commit(&mut self) -> PgQueueResult<()> {
+        let mut txn_guard = self.shared_txn.lock().await;
+        txn_guard.commit().await
+    }
+}
+
+/// A shared transaction. Exists as a named type so that it can have some helper methods.
+#[derive(Debug)]
+struct SharedTxn<'c> {
+    /// The actual transaction object. If a transaction error occurs (e.g. the connection drops)
+    /// then the transaction should be dropped so that other jobs don't each try to do DB work that
+    /// is bound to fail.
+    raw_txn: Option<sqlx::Transaction<'c, sqlx::postgres::Postgres>>,
+}
+
+impl SharedTxn<'_> {
+    /// Commits and then drops the transaction.
+    pub async fn commit(&mut self) -> PgQueueResult<()> {
+        if self.raw_txn.is_none() {
+            return Err(PgQueueError::TransactionAlreadyClosedError);
+        }
+
+        let txn = self.raw_txn.take().unwrap();
+        txn.commit().await.map_err(|e| PgQueueError::QueryError {
+            command: "COMMIT".to_owned(),
+            error: e,
+        })?;
+
+        Ok(())
+    }
+
+    /// See `raw_txn` above, this should be called when a transaction error occurs (e.g. the
+    /// connection drops) so that other jobs don't each try to do DB work that is bound to fail.
+    fn drop_txn(&mut self) {
+        self.raw_txn = None;
+    }
+}
+
+impl SharedTxn<'_> {
```

Comment:
nit: while this is good engineering, we are in the "if I poke at this struct it might break in ways I don't yet understand" red zone, so I'd be fine dropping this.

sqlx transactions have [an `open` field set to false on `rollback()`](https://docs.rs/sqlx-core/0.7.3/src/sqlx_core/transaction.rs.html#92) so I'm assuming that the queries will fail-fast in that case. As `job.complete()` is simply an update query, I don't think there's a lot of value in handling that failure mode ourselves just yet.
I'd just have jobs call `tx.rollback()` on unrecoverable error, and let the rest propagate.

Reply:
Good call. Honestly, `SharedTxn` kind of fell out of my previous attempts to do the `commit` via `Drop`, and it does seem confusing and vestigial now with fresh eyes.

Reply:
Removed it, I think it makes things a lot easier to follow.